### ISSUE TITLE
Renewal failed: ngx.timer: Too many pending timers

### ISSUE THREAD

**Issue by bryanus** [GUI/lua-resty-auto-ssl#152]
I'm trying to debug why a few domain certificates have not renewed as they are less than 30 days to expiration, and other domains on the same server have renewed successfully.

I have looked at my error logs, but I can't find the domains in question, nor do I see any blocking-type errors that would prevent certificates from being issued/re-issued.

Is it possible to renew a certificate manually, so I can watch my error logs when it runs to see if I can gain any useful information from it?

TIA.

**Comment by bryanus** [comment]
With a lot of digging around in the open/closed issues, I found the command that appears to run auto-ssl on a specified domain:

````
sudo -u ec2-user env HOOK_SECRET=XXXxxx HOOK_SERVER_PORT=8999 /usr/local/openresty/luajit/bin/resty-auto-ssl/dehydrated --cron --no-lock --domain www.mydomain.com --challenge http-01 --config /etc/resty-auto-ssl/letsencrypt/config --hook /usr/local/openresty/luajit/bin/resty-auto-ssl/letsencrypt_hooks
````
which returned the following output:

````
# INFO: Using main config file /etc/resty-auto-ssl/letsencrypt/config
Processing www.mydomain.com
 + Checking domain name(s) of existing cert... unchanged.
 + Checking expire date of existing cert...
 + Valid till Oct 10 06:49:12 2018 GMT (Less than 30 days). Renewing!
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting authorization for www.marcosotelo.com...
 + 1 pending challenge(s)
 + Deploying challenge tokens...
 + Responding to challenge for www.marcosotelo.com authorization...
 + Challenge is valid!
 + Requesting certificate...
 + Checking certificate...
 + Done!
 + Creating fullchain.pem...
 + Using cached chain!
 + Done!
````

Hmm. So it renewed without any issues when run manually. But why isn't it being renewed by cron? I have several hundred other domains that were renewed without any issues, but have discovered a handful that didn't renew for some reason. I have not defined an renew_check_interval setting in my config, so the default is to run daily. 

Not sure what to check for. The only thing I can think of is to change the renew_check_interval interval check to be more often to see if anything pops up in the logs...

Where is the cron to run the daily renewal checks located? Maybe I can run that manually and see what happens...

**Comment by bryanus** [comment]
OK, so I'm getting closer...

I resorted to specifying the renew_check_interval in my config and setting it to an hour. I also changed my log level to 'info' to capture more...info.

Well, the auto_renew kicked off and so far, about a hundred+ domains are currently renewing certs.

It makes me think that the cron to run the check daily has not been running since my first batch of domains reissued a few days ago.

I've now left the renew_check_interval in the config, but changed it back to daily (86400), and will review my logs tomorrow to see if the checks ran.

Can anyone give me insight into when the renew cron runs? Is it midnight?

I don't see any related crons in /etc/cron.daily, so how/where is it getting kicked off?

Is there a dehydrated command to run the renewal on ALL certs? I can only seem to run it on a single domain at a time. (If I could see the command the cron is running, that would probably work, too.)

TIA.

**Comment by bryanus** [comment]
I think I found the issue:

````
[error] 14813#14813: *14297181 [lua] renewal.lua:189: auto-ssl: failed to create timer: too many pending timers, context: ngx.timer
````
I see this error across several of my error log files. Any idea what would cause this error to occur? And why didn't I get the error when I ran the renewal today? Do they clear themselves out?

Can anyone tell me

**Comment by bryanus** [comment]
Just to add an update on this issue...

I've found that the renewal checks seem to kick off at 06:25:02AM UTC (still can't find this cron, btw), but every 2 days and not every day as specified by the renew_check_interval in my config (set to 86400, i.e. 1 day). Not sure why that is, but at least it has been running. Also, it seems that the renewal checks/process is kicked off by restarting auto-ssl, rather than running a command to process the domains directly. Restarting openresty in turn creates a new log file, which I can see occurs exactly at 06:25:02AM UTC; the old file ends at that time, and the new one starts.

Also, I am not seeing the 'too many pending timers' error anymore, so hopefully that won't occur again.

At this point, all I can do now is wait another 60 days or so and see if my domain certs are reissued and continue to monitor logs in the meantime.

Closing for now...
